### PR TITLE
[LT] Sync LTC branch changes on torch/csrc/lazy/core

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69012

Summary: Some changes to torch/csrc/lazy/core were done on the
lazy_tensor_staging branch (https://github.com/pytorch/pytorch/pull/68427).
Merge those back into the trunk.

Differential Revision: [D32708696](https://our.internmc.facebook.com/intern/diff/D32708696)

### CODE REVIEW

File: torch/csrc/lazy/ts_backend/ts_lowering_context.h [hunk 1]
```diff
@@ -74,16 +74,48 @@ class TORCH_API TSLoweringContext : public LoweringContext {
...
-  torch::jit::Value* GetOutputOp(const Output& output);
+  torch::jit::Value* GetOutputOp(const Output& output) {
+    auto it = emitted_outputs_.find(output);
+    if (it == emitted_outputs_.end()) {
+      auto post_order = Util::ComputePostOrder(output.node, &emit_status_);
+      for (auto node : post_order) {
+        bool ok = lowering_->Lower(node);
+        CHECK(ok) << "Failed to lower: " << *node;
+      }
+      // At this point the outpout better be present, otherwise there is an issue
```
Comment (wconstab) [inline torch/csrc/lazy/ts_backend/ts_lowering_context.h:111]: output

File: torch/csrc/lazy/ts_backend/ts_node_lowering.h [hunk 2]
```diff
@@ -3,14 +3,14 @@
 #include <torch/csrc/api/include/torch/jit.h>
 #include <torch/csrc/lazy/backend/lowering_context.h>
 
-namespace torch_lazy_tensors {
```
Comment (wconstab) [inline torch/csrc/lazy/ts_backend/ts_node_lowering.h:6]: oops, i landed it this way?

File: torch/csrc/lazy/backend/backend_interface.h [hunk 3]
```diff
@@ -119,7 +119,7 @@ class TORCH_API BackendRegistrar {
   BackendRegistrar(const BackendImplInterface* backend_impl_interface);
 };
 
-inline const BackendImplInterface* TORCH_API getBackend() {
+static inline const BackendImplInterface* getBackend() {
```
Comment (wconstab) [inline torch/csrc/lazy/backend/backend_interface.h:122]: we don't need TORCH_API macro here?

**Review by wconstab — APPROVED** [review]

File: torch/csrc/lazy/backend/backend_interface.h (hunk 3 above)
Comment (desertfire) [reply torch/csrc/lazy/backend/backend_interface.h:122]: For a static inline function in a header file, we don't need it. I removed it according to a lint complain.

File: torch/csrc/lazy/ts_backend/ts_node_lowering.h (hunk 2 above)
Comment (desertfire) [reply torch/csrc/lazy/ts_backend/ts_node_lowering.h:6]: Yep

File: torch/csrc/lazy/core/hash.h [hunk 4]
```diff
@@ -137,6 +181,11 @@ static inline hash_t Hash(const hash_t& value) {
   return value;
 }
 
+template <typename T>
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/hash.h:184]: Aha, I have a WIP to add this to the master. Feel free to do so and I probably just add the unit tests.

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32708696).

**Pull request closed without merging**